File: scratch/benchmark_4gb.py
import os
import time
import sys
import json
import urllib.request
import urllib.parse

SERVER_URL = "http://127.0.0.1:8000"
FILE_SIZE_GB = 4
FILE_SIZE_BYTES = FILE_SIZE_GB * 1024 * 1024 * 1024
TEMP_FILE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "speedtest_4gb.tmp")

def get_active_session_token():
    try:
        url = f"{SERVER_URL}/api/benchmark/token"
        req = urllib.request.Request(url)
        with urllib.request.urlopen(req) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            return data["session_token"]
    except Exception as e:
        print(f"Failed to fetch session token from server: {e}")
        print("Please make sure the server is running on port 8000.")
        sys.exit(1)

def create_4gb_file():
    print(f"Creating a 4GB dummy file for speed testing at: {TEMP_FILE_PATH}")
    buffer_size = 16 * 1024 * 1024  # 16MB
    data = b"a" * buffer_size
    written = 0
    start = time.time()
    
    with open(TEMP_FILE_PATH, "wb") as f:
        while written < FILE_SIZE_BYTES:
            to_write = min(buffer_size, FILE_SIZE_BYTES - written)
            f.write(data[:to_write])
            written += to_write
            if time.time() - start > 1:
                pct = (written / FILE_SIZE_BYTES) * 100
                print(f"Writing dummy file: {pct:.1f}%...")
                start = time.time()
                
    print("Successfully created 4GB dummy file.")

def run_upload_test(sid):
    print("\n--- STARTING 4GB UPLOAD SPEED TEST ---")
    url = f"{SERVER_URL}/api/files/upload_raw?filename=speedtest_4gb.tmp&sid={urllib.parse.quote(sid)}"
    print("Uploading 4GB payload to local server...")
    start_time = time.time()
    
    try:
        import httpx
        
        def generator():
            bytes_sent = 0
            last_report = time.time()
            with open(TEMP_FILE_PATH, "rb") as f:
                while True:
                    chunk = f.read(4 * 1024 * 1024) # 4MB chunks
                    if not chunk:
                        break
                    bytes_sent += len(chunk)
                    now = time.time()
                    if now - last_report >= 1.0:
                        pct = (bytes_sent / FILE_SIZE_BYTES) * 100
                        speed = (bytes_sent / (now - start_time) / 1048576)
                        print(f"Uploading: {pct:.1f}% | Current Speed: {speed:.2f} MB/s")
                        last_report = now
                    yield chunk

        with httpx.Client(timeout=600.0) as client:
            r = client.post(url, content=generator(), headers={"Content-Type": "application/octet-stream"})
            print(f"Server response: {r.json()}")
            duration = time.time() - start_time
            avg_speed = (FILE_SIZE_BYTES / duration / 1048576)
            print(f"Average Upload Speed: {avg_speed:.2f} MB/s (Time taken: {duration:.2f}s)")
            return avg_speed
    except ImportError:
        print("Please install httpx to run this benchmark: pip install httpx")
        return 0

def run_download_test(sid):
    print("\n--- STARTING 4GB DOWNLOAD SPEED TEST ---")
    url = f"{SERVER_URL}/api/files/download/speedtest_4gb.tmp?sid={urllib.parse.quote(sid)}"
    start_time = time.time()
    bytes_received = 0
    last_report = time.time()
    
    req = urllib.request.Request(url, headers={"User-Agent": "LANpad Benchmark"})
    with urllib.request.urlopen(req) as response:
        while True:
            chunk = response.read(4 * 1024 * 1024) # 4MB chunks
            if not chunk:
                break
            bytes_received += len(chunk)
            now = time.time()
            if now - last_report >= 1.0:
                pct = (bytes_received / FILE_SIZE_BYTES) * 100
                speed = (bytes_received / (now - start_time) / 1048576)
                print(f"Downloading: {pct:.1f}% | Current Speed: {speed:.2f} MB/s")
                last_report = now
                
    duration = time.time() - start_time
    avg_speed = (bytes_received / duration / 1048576)
    print(f"Average Download Speed: {avg_speed:.2f} MB/s (Time taken: {duration:.2f}s)")
    return avg_speed

if __name__ == "__main__":
    sid = get_active_session_token()
    
    if not os.path.exists(TEMP_FILE_PATH):
        create_4gb_file()
    
    try:
        up = run_upload_test(sid)
        dl = run_download_test(sid)
        
        print("\n======================================")
        print("         BENCHMARK RESULTS            ")
        print("======================================")
        print(f"Upload Speed:   {up:.2f} MB/s")
        print(f"Download Speed: {dl:.2f} MB/s")
        print("======================================\n")
    finally:
        if os.path.exists(TEMP_FILE_PATH):
            os.remove(TEMP_FILE_PATH)
        # Try to delete the uploaded test file from sharing dir
        try:
            url = f"{SERVER_URL}/api/files/delete/speedtest_4gb.tmp?sid={urllib.parse.quote(sid)}"
            req = urllib.request.Request(url, method="DELETE")
            urllib.request.urlopen(req)
        except Exception:
            pass

File: scratch/benchmark_4gb.js
const fs = require('fs');
const http = require('http');
const path = require('path');

// Configuration
const SERVER_URL = 'http://127.0.0.1:8000';
const SID = 'test_session_token'; // Replace with active SESSION_TOKEN if needed
const FILE_SIZE_GB = 4;
const FILE_SIZE_BYTES = FILE_SIZE_GB * 1024 * 1024 * 1024;
const TEMP_FILE_PATH = path.join(__dirname, 'speedtest_4gb.tmp');

function create4GBFile() {
    console.log(`Creating a 4GB dummy file for speed testing at: ${TEMP_FILE_PATH}`);
    const bufferSize = 16 * 1024 * 1024; // 16MB write buffer
    const buffer = Buffer.alloc(bufferSize, 'a');
    const fd = fs.openSync(TEMP_FILE_PATH, 'w');
    let written = 0;
    
    const start = Date.now();
    while (written < FILE_SIZE_BYTES) {
        const toWrite = Math.min(bufferSize, FILE_SIZE_BYTES - written);
        fs.writeSync(fd, buffer, 0, toWrite);
        written += toWrite;
        
        if (Date.now() - start > 1000) {
            const pct = ((written / FILE_SIZE_BYTES) * 100).toFixed(1);
            console.log(`Writing dummy file: ${pct}%...`);
        }
    }
    fs.closeSync(fd);
    console.log(`Successfully created 4GB dummy file in ${((Date.now() - start) / 1000).toFixed(2)}s`);
}

async function runUploadTest() {
    console.log('\n--- STARTING 4GB UPLOAD SPEED TEST ---');
    const fileStream = fs.createReadStream(TEMP_FILE_PATH, { highWaterMark: 4 * 1024 * 1024 }); // 4MB stream chunks
    
    return new Promise((resolve, reject) => {
        const url = `${SERVER_URL}/api/files/upload_raw?filename=speedtest_4gb.tmp&sid=${encodeURIComponent(SID)}`;
        const parsedUrl = new URL(url);
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port,
            path: parsedUrl.pathname + parsedUrl.search,
            method: 'POST',
            headers: {
                'Content-Type': 'application/octet-stream',
                'Content-Length': FILE_SIZE_BYTES
            }
        };

        const startTime = Date.now();
        let bytesSent = 0;
        let lastReportTime = Date.now();

        const req = http.request(options, (res) => {
            let responseData = '';
            res.on('data', (chunk) => responseData += chunk);
            res.on('end', () => {
                const totalDuration = (Date.now() - startTime) / 1000;
                const avgSpeed = (FILE_SIZE_BYTES / totalDuration / 1048576).toFixed(2);
                console.log(`\nUpload complete. Server response: ${responseData}`);
                console.log(`Average Upload Speed: ${avgSpeed} MB/s (Time taken: ${totalDuration.toFixed(2)}s)`);
                resolve(avgSpeed);
            });
        });

        req.on('error', (err) => {
            console.error('Upload Request Error:', err);
            reject(err);
        });

        // Monitor progress
        fileStream.on('data', (chunk) => {
            bytesSent += chunk.length;
            const now = Date.now();
            if (now - lastReportTime >= 1000) {
                const elapsed = (now - startTime) / 1000;
                const speed = (bytesSent / elapsed / 1048576).toFixed(2);
                const pct = ((bytesSent / FILE_SIZE_BYTES) * 100).toFixed(1);
                console.log(`Uploading: ${pct}% | Current Speed: ${speed} MB/s`);
                lastReportTime = now;
            }
        });

        fileStream.pipe(req);
    });
}

async function runDownloadTest() {
    console.log('\n--- STARTING 4GB DOWNLOAD SPEED TEST ---');
    return new Promise((resolve, reject) => {
        const url = `${SERVER_URL}/api/files/download/speedtest_4gb.tmp?sid=${encodeURIComponent(SID)}`;
        const parsedUrl = new URL(url);
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port,
            path: parsedUrl.pathname + parsedUrl.search,
            method: 'GET'
        };

        const startTime = Date.now();
        let bytesReceived = 0;
        let lastReportTime = Date.now();

        const req = http.request(options, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(`Server returned code ${res.statusCode}`));
                return;
            }

            res.on('data', (chunk) => {
                bytesReceived += chunk.length;
                const now = Date.now();
                if (now - lastReportTime >= 1000) {
                    const elapsed = (now - startTime) / 1000;
                    const speed = (bytesReceived / elapsed / 1048576).toFixed(2);
                    const pct = ((bytesReceived / FILE_SIZE_BYTES) * 100).toFixed(1);
                    console.log(`Downloading: ${pct}% | Current Speed: ${speed} MB/s`);
                    lastReportTime = now;
                }
            });

            res.on('end', () => {
                const totalDuration = (Date.now() - startTime) / 1000;
                const avgSpeed = (bytesReceived / totalDuration / 1048576).toFixed(2);
                console.log(`\nDownload complete.`);
                console.log(`Average Download Speed: ${avgSpeed} MB/s (Time taken: ${totalDuration.toFixed(2)}s)`);
                resolve(avgSpeed);
            });
        });

        req.on('error', (err) => {
            console.error('Download Request Error:', err);
            reject(err);
        });

        req.end();
    });
}

async function main() {
    try {
        // Step 1: Create 4GB file
        if (!fs.existsSync(TEMP_FILE_PATH)) {
            create4GBFile();
        } else {
            console.log('4GB dummy file already exists. Skipping creation.');
        }

        // Step 2: Run Upload Speed Test
        const uploadSpeed = await runUploadTest();

        // Step 3: Run Download Speed Test
        const downloadSpeed = await runDownloadTest();

        console.log('\n======================================');
        console.log('         SPEED TEST COMPLETED         ');
        console.log('======================================');
        console.log(`Laptop Local Upload speed:   ${uploadSpeed} MB/s`);
        console.log(`Laptop Local Download speed: ${downloadSpeed} MB/s`);
        console.log('--------------------------------------');
        console.log('For mobile: Use Private/Incognito browser on');
        console.log('your phone connected to LAN, run the test card,');
        console.log('and check real-time speed feedback there!');
        console.log('======================================\n');

        // Cleanup temporary files
        try {
            fs.unlinkSync(TEMP_FILE_PATH);
            console.log('Cleaned up local dummy file.');
        } catch (e) {}

    } catch (err) {
        console.error('Speed test execution failed:', err);
    }
}

main();
